[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/nickdlc/CSc448-Projects/blob/main/Assignment4/Iris-Analysis.ipynb)

In [912]:
from sklearn.datasets import load_iris

import numpy as np
import pandas as pd

# Load the Dataset

In [913]:
# Load the iris dataset and show the first 15 rows
iris = load_iris(as_frame=True)
df = iris['frame']
df.head(15)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

    target  
0        0  
1        0  
2        0  
3        0  
4        0  
5        0  
6        0  
7        0  
8        0  
9        0  
10       0  
11       0  
12       0  
13       0  
14       0

In [914]:
# Show a summary of the dataset
df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)      target  
count        150.000000  150.000000  
mean           1.199333    1.000000  
std            0.762238    0.819232  
min            0.100000    0.000000  
25%            0.300000    0.000000  
50%            1.300000    1.000000  
75%            1.800000    2.000000  
max            2.500000    2.000000

In [915]:
# Check for null values
df.isna().sum()

sepal length (cm)    0
sepal width (cm)     0
petal length (cm)    0
petal width (cm)     0
target               0
dtype: int64

In [916]:
# Get unique values of 'target' column
df['target'].unique()

array([0, 1, 2])

## About the Dataset

The dataset contains 150 flowers and records four different characteristics for each flower:
* Sepal length (cm)
* Sepal width (cm)
* Petal length (cm)
* Petal width (cm)

Moreover, the dataset contains a target column which has three possible values from the set {0,1,2}. These three distinct possibilities likely mean that there are three different flower species in the dataset, and these flower properties are the features used to classify each flower.

# Split the Dataset

In [917]:
from sklearn.model_selection import train_test_split

# Divide the dataset into features and label(s)
X = df.drop(['target'], axis=1)
y = df['target']

# Split the data into train (90%) and test (10%) sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=22)

# Split the data into train (67%) and test (33%) sets
# X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=22)

# Logistic Regression

## Training the Logistic Regression Model

In [918]:
from sklearn.linear_model import LogisticRegression

# Train the model using our training data
logreg = LogisticRegression(random_state=22, max_iter=200) # set max_iter to reach convergence and avoid warning
logreg.fit(X_train, y_train)

# Predict the output for each flower in the test set
y_pred = logreg.predict(X_test)

## Predicting Probabilities for Each Class for Sample Data Points

In [919]:
# Create a new point and predict the probabilites for each classification
X0 = [[15.8, 8.8, 13.8, 5], [2.15, 1, 0.5, 0.05], [5.8, 3, 4.35, 1.3]]
X0_pred = pd.DataFrame(X0, columns=X_test.columns)
y0_probs = logreg.predict_proba(X0_pred)

for i in range(len(y0_probs)):
    print('Point X' + str(i))
    for j in range(len(y0_probs[i])):
        print('Probability of class', str(j) + ':', y0_probs[i][j])
    print()

Point X0
Probability of class 0: 9.886265607331051e-25
Probability of class 1: 8.855563560036231e-13
Probability of class 2: 0.9999999999991145

Point X1
Probability of class 0: 0.9953922955320623
Probability of class 1: 0.004607701365240459
Probability of class 2: 3.102697239824551e-09

Point X2
Probability of class 0: 0.013854012075532204
Probability of class 1: 0.8984185571453003
Probability of class 2: 0.08772743077916749



In [920]:
# Verify that the model chooses the class with the highest probability
y0_pred = logreg.predict(X0_pred)

for i in range(len(y0_pred)):
    print('Given a new point X' + str(i), 'the model classifies it as iris', y0_pred[i])

Given a new point X0 the model classifies it as iris 2
Given a new point X1 the model classifies it as iris 0
Given a new point X2 the model classifies it as iris 1


## Performance Analysis

In [921]:
# Display the model's score
logreg_score = logreg.score(X_test, y_test)
# logreg_score = logreg.score(X0_pred, y0_pred)
print('The model\'s score is:', logreg_score)

The model's score is: 1.0


The logistic regression model has a score of 1, and this means that it can classify a flower as one of the three iris species with about 100% accuracy from the testing set or an entirely new point $X_0$.

When using a training set size of 67% and a test set size of 33%, this model has a score of 0.96, meaning it can properly classify the iris species with about 96% accuracy.

## Results

In [922]:
logreg_coefs = logreg.coef_
logreg_intercept = logreg.intercept_

print('The coefficients are:', logreg_coefs)
print('The intercepts are:', logreg_intercept)

The coefficients are: [[-0.41663174  0.95122844 -2.44264802 -1.04324609]
 [ 0.4532748  -0.32750687 -0.1992296  -0.80915027]
 [-0.03664306 -0.62372158  2.64187762  1.85239636]]
The intercepts are: [  9.53861062   2.43822503 -11.97683565]


# Support Vector Machine

## Training the Support Vector Model

In [923]:
from sklearn.svm import SVC

svclf = SVC(random_state=22, probability=True)
svclf.fit(X_train, y_train)

y_pred_svm = svclf.predict(X_test)

## Predicting Probabilities for Each Class for Sample Data Points

In [924]:
# Use the new point X0 and predict the probabilities for each classification
y0_pred_svm = svclf.predict(X0_pred)
y0_probs_svm = svclf.predict_proba(X0_pred)

for i in range(len(y0_probs_svm)):
    print('Point X' + str(i))
    for j in range(len(y0_probs_svm[i])):
        print('Probability of class', str(j) + ':', y0_probs_svm[i][j])
    print()

Point X0
Probability of class 0: 0.3563909475965388
Probability of class 1: 0.2414443096329036
Probability of class 2: 0.4021647427705573

Point X1
Probability of class 0: 0.8143933713532433
Probability of class 1: 0.09096295563272797
Probability of class 2: 0.094643673014029

Point X2
Probability of class 0: 0.00899296022121033
Probability of class 1: 0.9636723559621025
Probability of class 2: 0.027334683816687192



In [925]:
# Verify that the model chooses the class with the highest probability
for i in range(len(y0_pred_svm)):
    print('Given a new point X' + str(i), 'the model classifies it as iris', y0_pred[i])

Given a new point X0 the model classifies it as iris 2
Given a new point X1 the model classifies it as iris 0
Given a new point X2 the model classifies it as iris 1


## Performance Analysis

In [926]:
# Display the model's score
svm_score = svclf.score(X_test, y_test)
# svm_score = svclf.score(X0_pred, y0_pred_svm)
print('The model\'s score is:', svm_score)

The model's score is: 0.9333333333333333


The support vector machine model has a score of 0.93, and this means that it can classify a flower as one of the three iris species with about 93% accuracy from the testing set or an entirely new point $X_0$.

When using a training set size of 67% and a test set size of 33%, this model has a score of 0.92, meaning it can properly classify the iris species with about 92% accuracy.

# Neural Network

## Training the Neural Network

In [927]:
from sklearn.neural_network import MLPClassifier

mlpclf = MLPClassifier(random_state=22, max_iter=600) # set max_iter to reach convergence and avoid warning
mlpclf.fit(X_train, y_train)

y_pred_mlp = mlpclf.predict(X_test)

## Predicting Probabilities for Each Class for Sample Data Points

In [928]:
# Use the new point X0 and predict the probabilities for each classification
y0_pred_mlp = mlpclf.predict(X0_pred)
y0_probs_mlp = mlpclf.predict_proba(X0_pred)

for i in range(len(y0_probs_mlp)):
    print('Point X' + str(i))
    for j in range(len(y0_probs_mlp[i])):
        print('Probability of class', str(j) + ':', y0_probs_mlp[i][j])
    print()

Point X0
Probability of class 0: 4.805637683626675e-16
Probability of class 1: 3.674451114418523e-08
Probability of class 2: 0.9999999632554883

Point X1
Probability of class 0: 0.9170162399708093
Probability of class 1: 0.08298188668576408
Probability of class 2: 1.8733434266134747e-06

Point X2
Probability of class 0: 0.0028154864224915663
Probability of class 1: 0.9877627893021205
Probability of class 2: 0.009421724275388016



In [929]:
# Verify that the model chooses the class with the highest probability
for i in range(len(y0_pred_mlp)):
    print('Given a new point X' + str(i), 'the model classifies it as iris', y0_pred_mlp[i])

Given a new point X0 the model classifies it as iris 2
Given a new point X1 the model classifies it as iris 0
Given a new point X2 the model classifies it as iris 1


## Performance Analysis

In [930]:
# Display the model's score
mlp_score = mlpclf.score(X_test, y_test)
# mlp_score = mlpclf.score(X0_pred, y0_pred_mlp)
print('The model\'s score is:', mlp_score)

The model's score is: 1.0


The neural network/MLP classifier model has a score of 1, and this means that it can classify a flower as one of the three iris species with 100% accuracy from the testing set or an entirely new point $X_0$.

When using a training set size of 67% and a test set size of 33%, this model has a score of 1, meaning it can properly classify the iris species with 100% accuracy.

## Experimenting with Different Options

### Activation Options

In [931]:
activation_options = ['identity', 'logistic', 'tanh'] # relu is the default

for option in activation_options:
    print('Using activation option: ' + option.upper())
    mlpclf = MLPClassifier(random_state=22, max_iter=750, activation=option)
    mlpclf.fit(X_train, y_train)
    mlp_score = mlpclf.score(X_test, y_test)
    print('The model\'s score is:', mlp_score, '\n')

Using activation option: IDENTITY
The model's score is: 1.0 

Using activation option: LOGISTIC
The model's score is: 1.0 

Using activation option: TANH
The model's score is: 1.0 



The activation options are the activation functions for the hidden layer(s) of the neural network. In this case, the activation option appears to have no bearing on the model's score as all four options yield a score of 1 when the train-test split is 90%-10%. This could likely be due to overfitting to the training data.

### Max Iteration Options

In [932]:
import warnings
from sklearn.exceptions import ConvergenceWarning

warnings.filterwarnings('ignore', category=ConvergenceWarning) # suppress convergence warnings

for i in range(50, 300, 50):
    print('Max iterations:', i)
    mlpclf = MLPClassifier(random_state=22, max_iter=i)
    mlpclf.fit(X_train, y_train)
    mlp_score = mlpclf.score(X_test, y_test)
    print('The model\'s score is:', mlp_score, '\n')

Max iterations: 50
The model's score is: 0.6 

Max iterations: 100
The model's score is: 0.9333333333333333 

Max iterations: 150
The model's score is: 1.0 

Max iterations: 200
The model's score is: 1.0 

Max iterations: 250
The model's score is: 1.0 



The max iterations option determines the maximum number of iterations to cap the model at if convergence is not reached. In this case, convergence is not reached until ~600 maximum iterations. The warnings were suppressed to avoid cluttering the results. We see that with lower iterations, the model's performance is worse. This makes sense since there are fewer iterations for the model to learn, but we still see the issue of the score reaching 1 long before the model converges.

### Overall Option Analysis

As we will see later on, it appears the model is overfitting to the data since the score is 1 in most cases. While they have perfect performance, there may still be issues in classifying an iris which is not similar to the data that appears in the data set.

# K-Nearest Neighbors

## Training the KNN Model

In [933]:
from sklearn.neighbors import KNeighborsClassifier

knnclf = KNeighborsClassifier()
knnclf.fit(X_train, y_train)

y_pred_knn = knnclf.predict(X_test)

## Predicting Probabilities for Each Class for Sample Data Points

In [934]:
# Use the new point X0 and predict the probabilities for each classification
y0_pred_knn = knnclf.predict(X0_pred)
y0_probs_knn = knnclf.predict_proba(X0_pred)

for i in range(len(y0_probs_knn)):
    print('Point X' + str(i))
    for j in range(len(y0_probs_knn[i])):
        print('Probability of class', str(j) + ':', y0_probs_knn[i][j])
    print()

Point X0
Probability of class 0: 0.0
Probability of class 1: 0.0
Probability of class 2: 1.0

Point X1
Probability of class 0: 1.0
Probability of class 1: 0.0
Probability of class 2: 0.0

Point X2
Probability of class 0: 0.0
Probability of class 1: 1.0
Probability of class 2: 0.0



In [935]:
# Verify that the model chooses the class with the highest probability
for i in range(len(y0_pred_mlp)):
    print('Given a new point X' + str(i), 'the model classifies it as iris', y0_pred_knn[i])

Given a new point X0 the model classifies it as iris 2
Given a new point X1 the model classifies it as iris 0
Given a new point X2 the model classifies it as iris 1


## Performance Analysis

In [936]:
# Display the model's score
knn_score = knnclf.score(X_test, y_test)
# knn_score = mlpclf.score(X0_pred, y0_pred_knn)
print('The model\'s score is:', knn_score)

The model's score is: 1.0


The KNN classifier model has a score of 1, and this means that it can classify a flower as one of the three iris species with about 100% accuracy from the testing set or an entirely new point $X_0$.

When using a training set size of 67% and a test set size of 33%, this model has a score of 0.96, meaning it can properly classify the iris species with about 96% accuracy.

# Conclusion

For the three sample data points, I deliberately chose $x_0, x_1,$ and $x_2$ to be double the max of each feature, half the min of each feature, and the 50th percentile of each feature to gain some intuition into how these features relate to the three classes. The results show that as the sepal length, sepal width, petal length, and petal width increase, the classification also "increases." In this context, "increases" refers to how the classifications, or iris species, appear to correspond to how large these 4 features are. This means that class 0 seems to correspond to smaller irises, class 1 seems to correspond to medium-sized irises, and class 2 seems to correspond to larger irises.

All of the models except for the SVM had scores of 1 while the SVM had a score of 0.93 when using a train set size of 90% and a test set size of 10%. This suggests that the models may have been overfit to the training data. By changing the train set and test set sizes to 67% and 33% respectively, the models had the following scores:
* Logistic regression -> 0.96
* Support vector machine -> 0.92
* Neural network -> 1
* K-nearest neighbors -> 0.96

While the scores are still quite high, these results seem to be better as it is unrealistic for most models to have a score of 1. This change can be seen as the models not overfitting to the training data as much. When the training set size is 90%, there are only 15 samples in the test set. This makes the probability of the models learning points similar to those in the testing set higher than if the training set size is 67%. The models will have a greater chance of properly classifying all of the test points as opposed to when the train set is smaller and the test set is larger as a result. However, even with this change, the models still come close to scores of 1. This reveals that many of the data points may be very similar to each other, allowing for the models to still accurately predict the class for a given iris.

Among the four models analyzed in this exercise, I believe the logistic regression model performed the best. Even though it did not have the highest score after modifying the training and testing split, analyzing the probabilities for the three sample data points $X_0, X_1,$ and $X_2$ shows that the logistic regression model had the highest probabilities for the "correct" classes for the original training and testing split, barring the KNN model. I believe this is better than the perfect probabilities of the KNN model for the same reasons as the scores; the logistic regression model had predictions close to 1, but also was more uncertain about $X_1$. This uncertainty shows that it is likely not as tightly fit to the original data as the KNN model. This is further supported by the scores coming from the modified training and testing split.

Seeing the perfect scores initially surprised me since it is unusual to come across models with scores that high, even while tweaking parameters and the data. After taking some time to think about how this could have happened, this result is less surprising.